### ISSUE TITLE
Dynamic Twisted Thread Pool

### ISSUE THREAD

**Issue by leth** [itamarst/crochet#70]
If you defer a number of things to thread each of which call EventualResult.wait() it is possible to fill the threadpool with blocked threads, causing hangs or timeouts.

It'd be great if EventualResult flagged the pool worker threads as blocked, allowing the pool to create new threads, and resolve the nested defer to threads.

**Comment by itamarst** [comment]
Hi,

Somehow I missed this when originally filed, sorry. Insofar as the thread pool implementation itself supports starting new threads as necessary, it seems like simply increasing the default maximum number of threads to a large number should solve most cases where this is a problem.

**Comment by leth** [comment]
You might sill block until timeouts if you somehow fill the pool with blocked threads. This POC has a few problems, namely it lacks locks around critical calls to change the pool size, but we have a fixed version which works for us. Feel free to close the issue. 

On 25 April 2015 18:53:34 BST, Itamar Turner-Trauring <notifications@github.com> wrote:
>Hi,
>
>Somehow I missed this when originally filed, sorry. Insofar as the
>thread pool implementation itself supports starting new threads as
>necessary, it seems like simply increasing the default maximum number
>of threads to a large number should solve most cases where this is a
>problem.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/itamarst/crochet/issues/70#issuecomment-96245917

-- 
Sent from my Phone. Please excuse my brevity.

**Comment by itamarst** [comment]
At a minimum before closing this I'm going to raise the max number from 20, which is far too low. It's probably not worth having a dynamically changing thread pool built-in because I imagine different people will have different preferences on how that would work.

**Comment by leth** [comment]
I was only talking about the case where every thread in the pool is blocked waiting on something which requires work in the pool, and thus none can make progress, apart from where timeouts free up pool space.

On 26 April 2015 16:48:25 BST, Itamar Turner-Trauring <notifications@github.com> wrote:
>At a minimum before closing this I'm going to raise the max number from
>20, which is far too low. It's probably not worth having a dynamically
>changing thread pool built-in because I imagine different people will
>have different preferences on how that would work.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/itamarst/crochet/issues/70#issuecomment-96402036

-- 
Sent from my Phone. Please excuse my brevity.

**Comment by itamarst** [comment]
Oh, interesting. Can you explain the situation where this happened?

If I remember correctly Crochet only uses the thread pool to push results back in to the calling threads, and that isn't blocking... So maybe the solution is for Crochet to have its own thread pool for that particular use case, and then it won't be impacted by other uses of the main reactor thread pool?

**Comment by itamarst** [comment]
Or... actually, Crochet doesn't use the reactor thread pool at all. So what was the situation where this was problem?

**Comment by leth** [comment]
The gist specifically identifies where a Pool thread tries to block on an EventualResult and raises the pool max size limit. I can't quite remember how to trigger that, but I'm sure you can come up with something!

**Comment by itamarst** [comment]
Ah, you were doing `deferToThread` yourself. Generally Crochet has no knowledge or expectation where the threads will come from, you could create your own thread pool yourself (`twisted.python.threadpool.ThreadPool`) or be running a WSGI app with its own thread pool or whatever. So certainly adjusting the thread pool is not Crochet's responsibility.

As far as adding a hook to `EventualResult`... there can be many other reasons a thread pool fills up, so that would be insufficient anyway. If a user wants dynamic adjustment they could e.g. start another thread that inspects the thread pool and its queue size on a regular basis. So I'm going to close this as out of scope for Crochet (but feel free to convince me otherwise if you feel that problem can't be solved outside of Crochet).

**Comment by leth** [comment]
Yeah, our application `defersToThread` a lot.
I think another part of our problem was that these blocked crochet tasks were starving twisted's deferred-to-thread DNS lookups, which was then timing out and breaking the async stuff.
Perhaps separating crochet's threadpool from twisted's threadpool might be an option for us, but we'd still have a problem if too many threads were blocking on their own threadpool.

Our current patches on the ThreadPool just fix the case where the pool would be fully blocked in this way, because then it has no chance of making progress.
We don't dynamically adjust it in any other situation, and that seems to be working okay for our workloads.

In theory you could have an arbitrarily long loop of threads blocked on threadpools, but then you probably have bigger problems in your application ;)
Fixing the simplest case of threads blocking against work their containing threadpool was enough.
